_posts/news/2021-01-12-announcing-scalajs-1.4.0.md

```diff
@@ -0,0 +1,188 @@
+
+If you need help with anything related to Scala.js, you may find our community [on Gitter](https://gitter.im/scala-js/scala-js) and [on Stack Overflow](https://stackoverflow.com/questions/tagged/scala.js).
+
+Bug reports can be filed [on GitHub](https://github.com/scala-js/scala-js/issues).
+
+## Release notes
+
+If upgrading from Scala.js 0.6.x, make sure to read [the release notes of Scala.js 1.0.0]({{ BASE_PATH }}/news/2020/02/25/announcing-scalajs-1.0.0/) first, as they contain a host of important information, including breaking changes.
+
+This is a **minor** release:
+
+* It is backward binary compatible with all earlier versions in the 1.x series: libraries compiled with 1.0.x through 1.3.x can be used with 1.4.0 without change.
+* It is *not* forward binary compatible with 1.3.x: libraries compiled with 1.4.0 cannot be used with 1.3.x or earlier.
+* It is *not* entirely backward source compatible: it is not guaranteed that a codebase will compile *as is* when upgrading from 1.3.x (in particular in the presence of `-Xfatal-warnings`).
+
+As a reminder, libraries compiled with 0.6.x cannot be used with Scala.js 1.x; they must be republished with 1.x first.
+
+## Fixes with compatibility concerns
+
+### `java.lang.Class.isAssignableFrom`
+
+[`java.lang.Class.isAssignableFrom` had some serious issues](https://github.com/scala-js/scala-js/issues/4328), which we fixed in Scala.js 1.4.0.
+It now completly follows its specification on the JVM.
+However, it is possible that some code was relying on the broken behavior, which was, in some cases, closer to how `isInstance` and `isInstanceOf` works on Scala.js.
+
+In particular, the following idiom, which we have seen before, is now broken more often than before:
+
+{% highlight scala %}
+def validate[T: ClassTag](x: Any): Option[T] = {
+  if (classTag[T].runtimeClass.isAssignableFrom(x.getClass()))
+    Some(x.asInstanceOf[T])
+  else
+    None
+}
+{% endhighlight %}
+
+Previously, if the `runtimeClass` was `java.lang.Integer`, and the value of `x` was the integer `5`, the above code would successfully return `Some(5)`.
+With the fixed `isAssignableFrom`, it will return `None` on Scala.js 1.4.0.
+
+This happens because `x.getClass()` returns the smallest numeric type that can hold the value `5`, i.e., `java.lang.Byte`.
+The previous implementation of `isAssignableFrom` would return `true` for `classOf[j.l.Integer].isAssignableFrom(classOf[j.l.Byte])`, but it now returns `false`.
+
+The above idiom can be fixed by advantageously replacing `isAssignableFrom` by `isInstance`:
+
+{% highlight scala %}
+def validate[T: ClassTag](x: Any): Option[T] = {
+  if (classTag[T].runtimeClass.isInstance(x))
+    Some(x.asInstanceOf[T])
+  else
+    None
+}
+{% endhighlight %}
+
+## Dynamic module loading
+
+In Scala.js 1.3.0, we had introduced [module splitting support]({{ BASE_PATH }}/doc/project/module.html#module-splitting).
+Scala.js 1.4.0 goes one step further with dynamic module loading.
```

Comment:
Might be worth mentioning that dynamic module loading works with any splitting mode and with any number of entry points.

Reply:
Good point. I added a sentence at the end of the paragraph.